src/Aspire.Hosting.Azure.AppContainers/AzureContainerAppsInfrastructure.cs

```diff
@@ -322,101 +321,69 @@ private void ProcessRoleAssignments()
+                        infra => AddRoleAssignmentsInfrastructure(infra, a, containerAppIdentityResource))
                     {
                         ProvisioningBuildOptions = provisioningOptions.ProvisioningBuildOptions,
                     };
 
+                    // existing resource role assignments need to be scoped to the resource's resource group
                     if (a.TryGetLastAnnotation<ExistingAzureResourceAnnotation>(out var existingAnnotation) &&
                         existingAnnotation.ResourceGroup is not null)
                     {
-                        existingResourceRoleAssignmentsResource.Scope = new(existingAnnotation.ResourceGroup);
+                        roleAssignmentsResource.Scope = new(existingAnnotation.ResourceGroup);
                     }
 
-                    existingResourceRoleAssignments.Add(existingResourceRoleAssignmentsResource);
+                    roleAssignments.Add(roleAssignmentsResource);
                 }
 
-                return existingResourceRoleAssignments;
+                return roleAssignments;
             }
 
-            private static void AddExistingResourceRoleAssignments(AzureResourceInfrastructure infra, AzureProvisioningResource existingResource, IEnumerable<RoleDefinition> roles, AzureProvisioningResource containerAppIdentityResource)
+            private void AddRoleAssignmentsInfrastructure(AzureResourceInfrastructure infra, AzureProvisioningResource azureResource, AzureProvisioningResource containerAppIdentityResource)
             {
-                var prefix = existingResource.GetBicepIdentifier();
-                var provisionable = existingResource.AddAsExistingResource(infra);
-
-                var appIdentityId = new BicepOutputReference("id", containerAppIdentityResource).AsProvisioningParameter(infra);
-                var appIdentityPrincipalId = new BicepOutputReference("principalId", containerAppIdentityResource).AsProvisioningParameter(infra);
-
-                foreach (var role in roles)
+                if (RoleAssignments.TryGetValue(azureResource, out var roles))
                 {
-                    var roleAssignment = CreateRoleAssignment(prefix, provisionable, role.Id, role.Name, appIdentityId, RoleManagementPrincipalType.ServicePrincipal, appIdentityPrincipalId);
-                    infra.Add(roleAssignment);
+                    var context = new AddRoleAssignmentsContext(
+                        infra,
+                        roles,
+                        new(() => RoleManagementPrincipalType.ServicePrincipal),
+                        new(() => new BicepOutputReference("principalId", containerAppIdentityResource).AsProvisioningParameter(infra, parameterName: AzureBicepResource.KnownParameters.PrincipalId)),
+                        new(() => new BicepOutputReference("principalName", containerAppIdentityResource).AsProvisioningParameter(infra, parameterName: AzureBicepResource.KnownParameters.PrincipalName)));
```

Comment:
Using Lazy in the public API is hella ugly. Would be cleaner if there was an `IRoleAssignmentsContext` and AddRoleAssignmentsContext as internal.

Reply:
Fixed in [Follow up on database role assignments (dotnet/aspire#8212)](https://github.com/dotnet/aspire/pull/8212)